File: programmers/문자열 압축.cpp
#include <string>
 
using namespace std;
 
int solution(string s) {
    int answer = s.length();
 
    if(s.length() == 1)
        return 1;
 
    for(int i = 1; i <= s.length() / 2; i++){ 
        int count = 1;
        string preStr = s.substr(0, i);
        string cmp, answerStr;
 
        for(int j = i; j < s.length(); j += i){ 
            cmp = s.substr(j, i);
 
            if(!(preStr.compare(cmp)))
                count++;
            else{
                if(count == 1){
                    answerStr += preStr;
                    preStr = cmp;
                }else{
                    answerStr = answerStr + to_string(count) + preStr;
                    preStr = cmp;
                    count = 1;
                }
            }
            
            if(j + i >= s.length()){
                if(count != 1){
                    answerStr = answerStr + to_string(count) + preStr;
                    break;
                }else{
                    answerStr = answerStr + s.substr(j);
                    break;
                }
            }
        }
        answer = (answer > answerStr.length()) ? answerStr.length() : answer;
    }
    return answer;
}

File: programmers/문자열 압축.java
class Solution {
    public int solution(String s) {
        int answer = 10000;
        if(s.length()==1)
            return 1;
        for(int i=0;i<=s.length()/2; i++){
            String temp=s.substring(0,i+1);
            String result="";
            int cnt=1;
            for(int j=i+1; j<s.length(); j+=i+1){
                if(result.length()>answer)
                    break;
                
                if(j+i+1<s.length()&&s.substring(j,j+i+1).equals(temp)){
                    cnt++;
                }
                else if(j+i+1<s.length()){
                    if(cnt>1){
                        result+=Integer.toString(cnt);
                    }
                    result+=temp;
                    temp=s.substring(j,j+i+1);
                    cnt=1;       
                }
                else{
                    if(!temp.equals(s.substring(j))){
                        result+=s.substring(j);
                    }
                    else{
                        cnt++;
                    }
                    if(cnt>1){
                        result+=Integer.toString(cnt);
                    }
                    result+=temp;
                }
            }
            if(result.length()<answer){
                answer=result.length();
            }
        }
        return answer;
    }
}